staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go

```diff
@@ -1329,41 +1329,64 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
 			}
 		}
 
-		_, ok := original[k]
-		if !ok {
-			// If it's not in the original document, just take the patch value.
-			original[k] = patchV
-			continue
+		var patchKind reflect.Kind
+		_, foundInOriginal := original[k]
+		if patchV != nil {
+			patchType := reflect.TypeOf(patchV)
+			patchKind = patchType.Kind()
+			if foundInOriginal {
+				// If a field is existing both in original and patch, we'll compare their kind via
+				// reflection, and simply do an overwrite if they mismatch.
+				originalType := reflect.TypeOf(original[k])
+				if originalType != patchType {
+					original[k] = patchV
+					continue
+				}
+			}
 		}
 
-		originalType := reflect.TypeOf(original[k])
-		patchType := reflect.TypeOf(patchV)
-		if originalType != patchType {
-			original[k] = patchV
-			continue
-		}
 		// If they're both maps or lists, recurse into the value.
-		switch originalType.Kind() {
+		// NOTE: We're using patch's kind to do the "switch" b/c the original could be
+		// nil when replacing/deleting unexisting fields from the original map.
+		switch patchKind {
 		case reflect.Map:
 			subschema, patchMeta, err2 := schema.LookupPatchMetadataForStruct(k)
 			if err2 != nil {
+				// Handling unknown fields
+				if !foundInOriginal {
```

Comment:
If `err2` is not caused by an unknown field, I guess doing this blindly is not safe.
Ideally, `LookupPatchMetadataForStruct` should returned a typed error. And you only do this when that type of error is returned.